### DISCUSSION TITLE
Multiple instances on one server?

### DISCUSSION THREAD

**Discussion by cpbreton** [lloesche/valheim-server-docker#51] [Q&A]
Is it possible to run multiple instances of Valheim on one host/behind one public IP address? Or will Steam/Valheim only recognize the instance on the 2456-8 ports?

**Comment by InB4DevOps** [comment]
I read on reddit that some users have multiple servers running on the same IP. 
As long as you have enough cores & RAM I don't see an issue.

**Reply by cpbreton** [reply]
so just use different ports for each and Steam/Valheim clients will see them?

**Reply by InB4DevOps** [reply]
well, there's still the problem, that Steam / Valheim can only show ~5000 servers so you might need to connect via IP:port (new in latest version) or favourites list but other than that there shouldn't be a problem.

**Comment by Addyvan** [comment]
@cpbreton You should be fine as long as the ports don't overlap. Let me know how it goes.

**Reply by cpbreton** [reply]
Worked! Thank you!

**Reply by seventh88** [reply]
Can you share how it was achieved?

**Reply by lloesche** [reply]
@seventh88 see https://github.com/lloesche/valheim-server-docker/discussions/174

You would start a second container and map different ports.

E.g.
```
$ docker run -d \
    --name valheim-server1 \
    --cap-add=sys_nice \
    -p 2456-2457:2456-2457/udp \
    -v $HOME/valheim-server1/config:/config \
    -v $HOME/valheim-server1/data:/opt/valheim \
    lloesche/valheim-server

$ docker run -d \
    --name valheim-server2 \
    --cap-add=sys_nice \
    -p 2466-2467:2456-2457/udp \
    -v $HOME/valheim-server2/config:/config \
    -v $HOME/valheim-server2/data:/opt/valheim \
    lloesche/valheim-server

```
plus whatever else options you're using. Make sure to have at least 4 cores and around 16 GB RAM. Or 8 GB RAM and something like 20 GB swap.

Then you can connect to `valheim-server1` on port `2456` and to `valheim-server2` on port `2466`.

**Reply by seventh88** [reply]
Worked! Thank you!

**Reply by Mood93** [reply]
@lloesche Would we have to edit the below files as well? Or does the SERVER_PORT variable overwrite those?

[https://github.com/lloesche/valheim-server-docker/search?q=2456](url)
Dockerfile ln 144
valheim.service ln 19
valheim-status ln 163

The reason I ask is because when I use a different port, 'docker ps' shows 2456 is used to some degree.
<img width="335" alt="Screen Shot 2021-03-11 at 2 47 15 PM" src="https://user-images.githubusercontent.com/4393039/110853058-c3f17600-8278-11eb-84c6-ccb524fa6bb4.png">

Any recommendations? Thanks!

**Reply by lloesche** [reply]
@Mood93 there's no need to change any of those or even the `SERVER_PORT` variable to run any number of instances on the same system. All you need to do is tell Docker at startup to map the container ports to some other host ports.

Example:
```
-p 2456-2457:2456-2457/udp
```
means "Map container ports 2456 and 2457 to host ports 2456 and 2457 respectively".

But then
```
-p 2466-2467:2456-2457/udp
```
means "Map container ports 2456 and 2457 to host ports 2466 and 2467 respectively".

So you see, the ports inside the container can stay the same. Each container has its own networking and they don't conflict. Hence no need to change `SERVER_PORT`. It is only the ports those container ports get mapped to on the host that change.

For completeness:
The only reason I can think of why you would ever have to change `SERVER_PORT` is if you run Docker with host networking `--net host`. Which is not recommended. In this mode the container gets raw access to the hosts network interfaces and so if you would then run two containers with the same ports they would conflict.

**Reply by lloesche** [reply]
> The reason I ask is because when I use a different port, 'docker ps' shows 2456 is used to some degree

Can you describe how you run this container? Just `docker run`? If so what is the exact command you are using. Or `docker-compose up`? If so what does your compose file look like? Are you using the one from this repo?
